server-common/src/main/java/org/apache/kafka/server/group/share/DefaultStatePersister.java

```diff
@@ -174,6 +159,88 @@ public CompletableFuture<WriteShareGroupStateResult> writeState(WriteShareGroupS
+              Map<Integer, CompletableFuture<ReadShareGroupStateResponse>> partMap =
+                  futureMap.computeIfAbsent(topicData.topicId(), k -> new HashMap<>());
+              if (!partMap.containsKey(partitionData.partition())) {
+                partMap.put(partitionData.partition(), new CompletableFuture<>());
+              }
+              return stateManager.new ReadStateHandler(
+                  groupId,
+                  topicData.topicId(),
+                  partitionData.partition(),
+                  partitionData.leaderEpoch(),
+                  partMap.get(partitionData.partition()));
+            })
+            .collect(Collectors.toList()))
+        .flatMap(Collection::stream)
+        .collect(Collectors.toList());
+
+    for (PersisterStateManager.PersisterStateManagerHandler handler : handlers) {
+      stateManager.enqueue(handler);
+    }
+
+    // Combine all futures into a single CompletableFuture<Void>
+    CompletableFuture<Void> combinedFuture = CompletableFuture.allOf(futureMap.values().stream()
+        .flatMap(map -> map.values().stream()).toArray(CompletableFuture[]::new));
+
+    // Transform the combined CompletableFuture<Void> into CompletableFuture<ReadShareGroupStateResult>
+    return combinedFuture.thenApply(v -> {
+      List<TopicData<PartitionAllData>> topicsData = futureMap.keySet().stream()
+          .map(topicId -> {
+            List<PartitionAllData> partitionAllData = futureMap.get(topicId).entrySet().stream()
+                .map(partitionFuture -> {
+                  int partition = partitionFuture.getKey();
+                  CompletableFuture<ReadShareGroupStateResponse> future = partitionFuture.getValue();
+                  try {
+                    ReadShareGroupStateResponse partitionResponse = future.get();
+                    return partitionResponse.data().results().get(0).partitions().stream()
+                        .map(partitionResult -> PartitionFactory.newPartitionAllData(
+                            partitionResult.partition(),
+                            partitionResult.stateEpoch(),
+                            partitionResult.startOffset(),
+                            partitionResult.errorCode(),
+                            partitionResult.errorMessage(),
+                            partitionResult.stateBatches().stream().map(PersisterStateBatch::from).collect(Collectors.toList())
+                        ))
+                        .collect(Collectors.toList());
+                  } catch (InterruptedException | ExecutionException e) {
+                    log.error("Unexpected exception while getting data from share coordinator", e);
+                    return Collections.singletonList(PartitionFactory.newPartitionAllData(
+                        partition,
+                        -1,
+                        -1,
+                        Errors.forException(e).code(),
+                        "Error reading state from share coordinator",
```

Comment:
I think you could append the `Errors.forException(e).message()` to this string for extra problem determination information.